conn_browser.go

```diff
@@ -0,0 +1,311 @@
+	}
+	return nil
+}
+
+func (c *Conn) Write(b []byte) (n int, err error) {
+	defer func() {
+		if e := recover(); e != nil {
+			err = recoveredValueToError(e)
+		}
+	}()
+	if err := c.checkOpen(); err != nil {
+		return 0, err
+	}
+	uint8Array := js.Global().Get("Uint8Array").New(len(b))
+	for i, bi := range b {
+		uint8Array.SetIndex(i, bi)
+	}
+	c.Call("send", uint8Array.Get("buffer"))
+	return len(b), nil
+}
+
+// Close closes the connection. Only the first call to Close will receive the
+// close error, subsequent and concurrent calls will return nil.
+// This method is thread-safe.
+func (c *Conn) Close() error {
+	c.closeOnce.Do(func() {
+		c.Call("close")
+		c.signalClose()
+		c.releaseHandlers()
+	})
+	return nil
+}
+
+func (c *Conn) signalClose() {
+	c.closeSignalOnce.Do(func() {
+		close(c.closeSignal)
+	})
+}
+
+func (c *Conn) releaseHandlers() {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	if c.messageHandler != nil {
+		c.Call("removeEventListener", "message", *c.messageHandler)
+		c.messageHandler.Release()
+		c.messageHandler = nil
+	}
+	if c.closeHandler != nil {
+		c.Call("removeEventListener", "close", *c.closeHandler)
+		c.closeHandler.Release()
+		c.closeHandler = nil
+	}
+	if c.errorHandler != nil {
+		c.Call("removeEventListener", "error", *c.errorHandler)
+		c.errorHandler.Release()
+		c.errorHandler = nil
+	}
+}
+
+func (c *Conn) LocalAddr() net.Addr {
+	return c.localAddr
+}
+
+func getRemoteAddr(val js.Value) net.Addr {
+	rawURL := val.Get("url").String()
+	withoutPrefix := strings.TrimPrefix(rawURL, "ws://")
+	withoutSuffix := strings.TrimSuffix(withoutPrefix, "/")
+	return NewAddr(withoutSuffix)
+}
+
+func (c *Conn) RemoteAddr() net.Addr {
+	return c.remoteAddr
+}
+
+// TODO: Return os.ErrNoDeadline. For now we return nil because multiplexers
+// don't handle the error correctly.
+func (c *Conn) SetDeadline(t time.Time) error {
+	return nil
+}
+
+func (c *Conn) SetReadDeadline(t time.Time) error {
+	return nil
+}
+
+func (c *Conn) SetWriteDeadline(t time.Time) error {
+	return nil
+}
+
+func (c *Conn) setUpHandlers() {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	if c.messageHandler != nil {
+		// Message handlers already created. Nothing to do.
+		return
+	}
+	messageHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		arrayBuffer := args[0].Get("data")
+		data := arrayBufferToBytes(arrayBuffer)
+		c.currDataMut.Lock()
+		if _, err := c.currData.Write(data); err != nil {
+			c.currDataMut.Unlock()
+			return err
+		}
+		c.currDataMut.Unlock()
+
+		// Non-blocking signal
+		select {
+		case c.dataSignal <- struct{}{}:
+		default:
+		}
+
+		return nil
+	})
+	c.messageHandler = &messageHandler
+	c.Call("addEventListener", "message", messageHandler)
+
+	closeHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		go func() {
+			c.signalClose()
+			c.mut.Lock()
+			// Store the error in c.firstErr. It will be returned by Read later on.
+			c.firstErr = errorEventToError(args[0])
```

Comment:
I see we're taking the lock when setting this but not when reading it. We should either:

1. Take the lock when reading it.
2. Guard it with the channel:
  1. Set it within `signalClose(err)`.
  2. Set it _before_ closing.
  3. Only read it after observing that the close-signaling channel is actually closed.